File: Week 3 & 4/04. Topics/ros2_ws/src/pkg_cpp_py/src/transistor_radio.cpp
#include "rclcpp/rclcpp.hpp"
#include "example_interfaces/msg/string.hpp"

using std::placeholders::_1;

class TransistorRadio : public rclcpp::Node
{
public:
    TransistorRadio() : Node("transistor_radio")
    {
        subscriber_ = this->create_subscription<example_interfaces::msg::String>("robot_news", 10, std::bind(&TransistorRadio::robot_news_callback, this,_1));
    }

private:
    void robot_news_callback(const example_interfaces::msg::String::SharedPtr msg)
    {
        RCLCPP_INFO(this->get_logger(), "%s", msg->data.c_str());
    }
    rclcpp::Subscription<example_interfaces::msg::String>::SharedPtr subscriber_;
};

int main(int argc, char **argv)
{
    rclcpp::init(argc, argv);
    rclcpp::spin(std::make_shared<TransistorRadio>());
    rclcpp::shutdown();
    return 0;
}

File: Week 3/04. Topics/ros2_ws/src/pkg_cpp_py/scripts/transistor_radio.py
#!/usr/bin/env python3

import rclpy
from rclpy.node import Node

from example_interfaces.msg import String

class TransitorRadio(Node):
    def __init__(self):
        super().__init__("transitor_radio")
        self.subscriber_ = self.create_subscription(String, "robot_news", robot_news_cb, 10)
        self.get_logger().info("Transitor radio has been tuned to a channel")

    def robot_news_cb(self, msg):
        self.get_logger().info(msg.data)

def main(args=None):
    rclpy.init(args=args)
    transitor_radio = TransitorRadio()
    rclpy.spin(transitor_radio)
    rclpy.shutdown()

if __name__ == '__main__':
    main()